Created SemanticLogger and associated Console and Mongo loggers

elimisteve opened the pull request:
Also created Payload data type for storing logging information, getting as close to the Ruby semantic_logger as possible.

elimisteve committed:
Semantic logging WIP. Added newlogger.go, which will be renamed to
logger.go once the former replaces the latter.

elimisteve committed:
Added Payload and Exception types

elimisteve committed:
Added two design proposals to semantic_logger.go

elimisteve committed:
Chose first SemanticLogger design (thanks @erikstmartin for your input)

elimisteve committed:
.gitignore now ignores Emacs and Vim temporary files

elimisteve committed:
Added MultiSemanticLogger type and necessary methods req'd to make it
implement the 'SemanticLogger' interface (eventually will be renamed
to just 'Logger')

elimisteve committed:
Implementing SemanticLogger now requires 2 methods, not 7

elimisteve committed:
Changed LogLevel to type int so they'd be sortable. Added
LessSevereThan() method to LogLevel's. Added custom String() method to
LogLevel's so printing a LogLevel looks like "TRACE" not "0".

elimisteve committed:
SemanticLogger's Log method now returns an error

elimisteve committed:
Added String() method to Exception type

elimisteve committed:
Added MongoSemanticLogger and methods req'd to implement
SemanticLogger interface

elimisteve committed:
Added Payload.UUID; clarified an error message

elimisteve committed:
ConsoleSemanticLogger and MongoSementicLogger now implement SemanticLogger

elimisteve committed:
Changed the types of Payload.PID and .Level

elimisteve committed:
MongoSemanticLogger.Log now panics when it should

elimisteve committed:
Got rid of Exception type (subsumed into Payload fields)

erikstmartin commented on semantic_logger.go:
```diff
@@ -0,0 +1,211 @@
+	switch level {
+	case TRACE, DEBUG, INFO, WARN, ERROR, FATAL:
+		for _, lgr := range ml {
+			// TODO: Decide what to do with returned `error` value
```
writing it to stdout is probably a safe bet

skelterjohn replied:
what returned error value?

elimisteve replied:
@skelterjohn The `Log` method returns an `error`

elimisteve replied:
@erikstmartin That's what it does in the latest version, which was also part of the Pull Request

erikstmartin commented:
We'll definitely need a FileLogger, at some point someone could do a syslog one if they wanted as well

erikstmartin commented on semantic_logger.go:
```diff
@@ -0,0 +1,211 @@
+type Exception struct {
+	Message   string   `json:"message"`
+	Backtrace []string `json:"backtrace"`
+}
```
We should really provide a way to help accomodate filling this backtrace, for example something like: https://gist.github.com/3899233

skelterjohn replied:
You can also use something like this: http://code.google.com/p/gonicetrace/ (I am the author, feel free to copy/paste code without citation if you want any of it)

elimisteve replied:
@erikstmartin This code, too, is included in the Pull Request. It's based upon some very helpful code that John wrote for the original logger.

skelterjohn commented on semantic_logger.go:
```diff
@@ -0,0 +1,211 @@
+	"time"
+)
+
+type Exception struct {
```
why do we need to consider exceptions at all?

elimisteve replied:
Two comments:
1. The idea is that within a `recover()` block we'd store the stacktrace info from panics. I'm calling them "exceptions" to maintain compatibility with [semantic_logger](https://github.com/ClarityServices/semantic_logger#semantic_logger).
2. The `Exception` type is removed in later commits; this info should be part of the `Payload` struct.

skelterjohn commented on semantic_logger.go:
```diff
@@ -0,0 +1,211 @@
+		backtrace)
+}
+
+type Payload struct {
```
this type should have godoc some comments

elimisteve replied:
Will add

elimisteve committed:
Categorized Payload fields into those that should be filled by
different parts of the program, and added comments specifying as
much. Made some fields unexported, unlike before.
